<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Dolev Michaelovich">
<meta name="author" content="Giulio Benedetti">

<title>Neurogenomics Project 1</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="fullcontent">

<div id="quarto-search-results"></div>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
<!-- margin-sidebar -->
    
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Neurogenomics Project 1</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Authors</div>
    <div class="quarto-title-meta-contents">
             <p>Dolev Michaelovich </p>
             <p>Giulio Benedetti </p>
          </div>
  </div>
    
  
    
  </div>
  

</header>

<p>This project was published on GitHub (https://riborings.github.io/neurogenomics/project1.html).</p>
<section id="part-1" class="level1">
<h1>Part 1</h1>
<p>Next-Generation Sequencing (NGS) entails several high-throughput techniques that generate a big amount of data. The main two approaches involve long-read sequencing with Nanopore technology and short-read sequencing with an Illumina machine. In the latter case, Polymerase Chain Reaction (PCR) is performed by the sequencer to amplify the library and quantify gene expression.</p>
<p>In the ideal situation, we would expect to have a doubling of each RNA strand in a linear fashion, so that each molecule is duplicated and the ratio between the expression levels of the different genes is maintained. However, this process introduces different types of errors due to factors such as the the composition of the sequence (for example high fraction of G or C) or the secondary structure, stochasticity and mutation that make it impossible to deduce the absolute gene expression, especially for low-copy genes.</p>
<p>In addition, the sequencing process involves breaking the mRNA sequence into several pieces because the machine cannot sequence a full-length genome. Thus, multiple variants may originate from the same sequence, so we may think there are several copies of the same gene, but in reality it is one broken copy. Therefore, we may infer a wrong expression level. Instead, sequencing data provides a compositional measure of gene expression, where every observation (or gene) is quantified in relation to the rest of the pool. Thus, it is essential to normalise reads from different samples when drawing between-sample comparisons.</p>
</section>
<section id="part-2" class="level1">
<h1>Part 2</h1>
<p>The package with the data of interest is imported.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Import package</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pasilla)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>RNA-Seq data from Drosophila melanogaster is loaded from the package.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load data</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>pasCts <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"pasilla_gene_counts.tsv"</span>,</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>                      <span class="at">package =</span> <span class="st">"pasilla"</span>, <span class="at">mustWork =</span> <span class="cn">TRUE</span>)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>pasAnno <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"pasilla_sample_annotation.csv"</span>,</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>                       <span class="at">package=</span><span class="st">"pasilla"</span>, <span class="at">mustWork =</span> <span class="cn">TRUE</span>)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>cts <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="fu">read.csv</span>(pasCts, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">row.names =</span> <span class="st">"gene_id"</span>))</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>coldata <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(pasAnno, <span class="at">row.names =</span> <span class="dv">1</span>)</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>coldata <span class="ot">&lt;-</span> coldata[ , <span class="fu">c</span>(<span class="st">"condition"</span>, <span class="st">"type"</span>)]</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(coldata) <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">"fb"</span>, <span class="st">""</span>, <span class="fu">rownames</span>(coldata))</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>cts <span class="ot">&lt;-</span> cts[, <span class="fu">rownames</span>(coldata)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The matrix <code>cts</code> contains the gene expression data in terms of the number of reads for 14599 genes from 7 samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># View head of counts assay</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(cts, <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>            treated1 treated2 treated3 untreated1 untreated2 untreated3
FBgn0000003        0        0        1          0          0          0
FBgn0000008      140       88       70         92        161         76
FBgn0000014        4        0        0          5          1          0
FBgn0000015        1        0        0          0          2          1
FBgn0000017     6205     3072     3334       4664       8714       3564
FBgn0000018      722      299      308        583        761        245
FBgn0000022        0        0        0          0          1          0
FBgn0000024       10        7        5         10         11          3
FBgn0000028        0        1        1          0          1          0
FBgn0000032     1698      696      757       1446       1713        615
            untreated4
FBgn0000003          0
FBgn0000008         70
FBgn0000014          0
FBgn0000015          2
FBgn0000017       3150
FBgn0000018        310
FBgn0000022          0
FBgn0000024          3
FBgn0000028          0
FBgn0000032        672</code></pre>
</div>
</div>
<p>Dimensions of the counts assay can be retrieved with the <code>dim</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Store assay dimensions</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>mat_dim <span class="ot">&lt;-</span> <span class="fu">dim</span>(cts)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="co"># View assay dimensions</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>mat_dim</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 14599     7</code></pre>
</div>
</div>
<p>Samples differ from one another in the library size. Therefore, read counts should be normalised.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># View raw library sizes</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colSums</span>(cts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  treated1   treated2   treated3 untreated1 untreated2 untreated3 untreated4 
  18670279    9571826   10343856   13972512   21911438    8358426    9841335 </code></pre>
</div>
</div>
<p>Normalisation is performed by dividing the reads from every sample by the corresponding library size and multiplying by the library size of the first sample.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Normalise all samples to library size of first sample</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>norm_cts <span class="ot">&lt;-</span> <span class="fu">apply</span>(cts, <span class="dv">2</span>, <span class="cf">function</span>(col) col <span class="sc">/</span> <span class="fu">sum</span>(col) <span class="sc">*</span> <span class="fu">sum</span>(cts[ , <span class="dv">1</span>]))</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Check normalised library sizes</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colSums</span>(norm_cts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  treated1   treated2   treated3 untreated1 untreated2 untreated3 untreated4 
  18670279   18670279   18670279   18670279   18670279   18670279   18670279 </code></pre>
</div>
</div>
</section>
<section id="part-3" class="level1">
<h1>Part 3</h1>
<p>A function to compute log means and variances for every feature in an assay is defined below, so that the same workflow can be applied both to untreated and treated samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Define function to compute feature-wise log means and vars</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>find_log_stats <span class="ot">&lt;-</span> <span class="cf">function</span>(assay, <span class="at">na.rm =</span> <span class="cn">FALSE</span>, <span class="at">pseudocount =</span> <span class="dv">1</span>) {</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Compute row means and vars</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>  gene_means <span class="ot">&lt;-</span> <span class="fu">rowMeans</span>(assay, <span class="at">na.rm =</span> na.rm)</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>  gene_vars <span class="ot">&lt;-</span> <span class="fu">rowVars</span>(assay, <span class="at">na.rm =</span> na.rm)</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Take log of means and vars with pseudocount</span></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>  gene_log_means <span class="ot">&lt;-</span> <span class="fu">log</span>(gene_means <span class="sc">+</span> pseudocount)</span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>  gene_log_vars <span class="ot">&lt;-</span> <span class="fu">log</span>(gene_vars <span class="sc">+</span> pseudocount)</span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Combine data into a data.frame</span></span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a>  df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Var =</span> gene_log_vars, <span class="at">Mean =</span> gene_log_means)</span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(df)</span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The log statistics of the genes from the untreated samples are computed.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset data to untreated samples</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>control_cts <span class="ot">&lt;-</span> norm_cts[ , <span class="fu">grep</span>(<span class="st">"^untreated"</span>, <span class="fu">colnames</span>(norm_cts))]</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Find log means and vars for genes in control samples</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>control_stats <span class="ot">&lt;-</span> <span class="fu">find_log_stats</span>(control_cts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The Poisson distribution fits the data relatively well for lower numbers of reads, but it creates bias for higher numbers of reads where the variance becomes larger than the mean. Such phenomenon is called <em>overdispersion</em>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualise fit of Poisson distribution to data</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(control_stats<span class="sc">$</span>Mean, control_stats<span class="sc">$</span>Var, <span class="at">xlab =</span> <span class="st">"Log Means"</span>, <span class="at">ylab =</span> <span class="st">"Log Vars"</span>)</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">a =</span> <span class="dv">0</span>, <span class="at">b =</span> <span class="dv">1</span>, <span class="at">col =</span> <span class="st">"red"</span>)</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topleft"</span>, <span class="st">"Poisson Fit"</span>, <span class="at">lty =</span> <span class="dv">1</span>, <span class="at">col =</span> <span class="st">"red"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="plots/unnamed-chunk-10-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The dispersion coefficient of the negative binomial distribution can be determined by fitting the corresponding mean-variance relationship to the data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Fit dispersion coefficient of negative binomial distribution to data</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>nls_fit <span class="ot">&lt;-</span> <span class="fu">nls</span>(Var <span class="sc">~</span> Mean <span class="sc">+</span> a <span class="sc">*</span> Mean<span class="sc">^</span><span class="dv">2</span>, <span class="at">data =</span> control_stats, <span class="at">start =</span> <span class="fu">list</span>(<span class="at">a =</span> <span class="dv">0</span>))</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(nls_fit)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Formula: Var ~ Mean + a * Mean^2

Parameters:
  Estimate Std. Error t value Pr(&gt;|t|)    
a 0.052201   0.000206   253.3   &lt;2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 0.8978 on 14598 degrees of freedom

Number of iterations to convergence: 1 
Achieved convergence tolerance: 1.408e-11</code></pre>
</div>
</div>
<p>Unlike the Poisson distribution, the negative binomial distribution fits well the data both for lower and higher numbers of reads.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Define var-mean relationship of negative binomial distribution</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>disper_coef <span class="ot">&lt;-</span> <span class="fu">summary</span>(nls_fit)<span class="sc">$</span>coefficients[<span class="st">"a"</span>, <span class="st">"Estimate"</span>]</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>neg_binom <span class="ot">&lt;-</span> <span class="cf">function</span>(x) x <span class="sc">+</span> disper_coef <span class="sc">*</span> x<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualise fit of negative binomial distribution to data</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(control_stats<span class="sc">$</span>Mean, control_stats<span class="sc">$</span>Var, <span class="at">xlab =</span> <span class="st">"Log Means"</span>, <span class="at">ylab =</span> <span class="st">"Log Vars"</span>)</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a><span class="fu">curve</span>(neg_binom, <span class="at">col =</span> <span class="st">"red"</span>, <span class="at">add =</span> <span class="cn">TRUE</span>)</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topleft"</span>, <span class="st">"Negative Binomial Fit"</span>, <span class="at">lty =</span> <span class="dv">1</span>, <span class="at">col =</span> <span class="st">"red"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="plots/unnamed-chunk-12-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The same pipeline is applied to the genes from the treated samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset data to treated samples</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>treated_cts <span class="ot">&lt;-</span> norm_cts[ , <span class="fu">grep</span>(<span class="st">"^treated"</span>, <span class="fu">colnames</span>(norm_cts))]</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Find log means and vars for genes in treated samples</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>treated_stats <span class="ot">&lt;-</span> <span class="fu">find_log_stats</span>(treated_cts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualise fit of Poisson distribution to data</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(control_stats<span class="sc">$</span>Mean, treated_stats<span class="sc">$</span>Var, <span class="at">xlab =</span> <span class="st">"Log Means"</span>, <span class="at">ylab =</span> <span class="st">"Log Vars"</span>)</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">a =</span> <span class="dv">0</span>, <span class="at">b =</span> <span class="dv">1</span>, <span class="at">col =</span> <span class="st">"red"</span>)</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topleft"</span>, <span class="st">"Poisson Fit"</span>, <span class="at">lty =</span> <span class="dv">1</span>, <span class="at">col =</span> <span class="st">"red"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="plots/unnamed-chunk-14-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Fit dispersion coefficient of negative binomial distribution to data</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>nls_fit <span class="ot">&lt;-</span> <span class="fu">nls</span>(Var <span class="sc">~</span> Mean <span class="sc">+</span> a <span class="sc">*</span> Mean<span class="sc">^</span><span class="dv">2</span>, <span class="at">data =</span> treated_stats, <span class="at">start =</span> <span class="fu">list</span>(<span class="at">a =</span> <span class="dv">0</span>))</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(nls_fit)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Formula: Var ~ Mean + a * Mean^2

Parameters:
   Estimate Std. Error t value Pr(&gt;|t|)    
a 0.0439520  0.0002798   157.1   &lt;2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 1.227 on 14598 degrees of freedom

Number of iterations to convergence: 1 
Achieved convergence tolerance: 7.173e-11</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Define var-mean relationship of negative binomial distribution</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>disper_coef <span class="ot">&lt;-</span> <span class="fu">summary</span>(nls_fit)<span class="sc">$</span>coefficients[<span class="st">"a"</span>, <span class="st">"Estimate"</span>]</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>neg_binom <span class="ot">&lt;-</span> <span class="cf">function</span>(x) x <span class="sc">+</span> disper_coef <span class="sc">*</span> x<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualise fit of negative binomial distribution to data</span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(treated_stats<span class="sc">$</span>Mean, treated_stats<span class="sc">$</span>Var, <span class="at">xlab =</span> <span class="st">"Log Means"</span>, <span class="at">ylab =</span> <span class="st">"Log Vars"</span>)</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a><span class="fu">curve</span>(neg_binom, <span class="at">col =</span> <span class="st">"red"</span>, <span class="at">add =</span> <span class="cn">TRUE</span>)</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topleft"</span>, <span class="st">"Negative Binomial Fit"</span>, <span class="at">lty =</span> <span class="dv">1</span>, <span class="at">col =</span> <span class="st">"red"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="plots/unnamed-chunk-16-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Judging by our results, we can estimate that the different untreated and treated samples in this experiment are biological repeats, because these changes are represented by a greater variability that corresponds to a negative binomial distribution and less to a Poisson distribution, this is due to the variability that exists between different animals/tissues. As we can see, our distribution also better fits a negative binomial distribution.</p>
</section>
<section id="part-4" class="level1">
<h1>Part 4</h1>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Transform normalised assay to log assay with pseudocount</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>log_cts <span class="ot">&lt;-</span> <span class="fu">log</span>(norm_cts <span class="sc">+</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualise log counts of treated vs untreated samples</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(log_cts[ , <span class="st">"untreated1"</span>], log_cts[ , <span class="st">"treated1"</span>],</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"Log Expression (untreated)"</span>, <span class="at">ylab =</span> <span class="st">"Log Expression (treated)"</span>)</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a><span class="fu">grid</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="plots/unnamed-chunk-18-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Identify gene that is expressed more in treated samples</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>control_cond <span class="ot">&lt;-</span> control_stats<span class="sc">$</span>Mean <span class="sc">&gt;</span> <span class="fl">8.5</span> <span class="sc">&amp;</span> control_stats<span class="sc">$</span>Mean <span class="sc">&lt;</span> <span class="fl">9.5</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>treated_cond <span class="ot">&lt;-</span> treated_stats<span class="sc">$</span>Mean <span class="sc">&gt;</span> <span class="dv">10</span> <span class="sc">&amp;</span> treated_stats<span class="sc">$</span>Mean <span class="sc">&lt;</span> <span class="dv">11</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>gene1 <span class="ot">&lt;-</span> <span class="fu">rownames</span>(norm_cts)[control_cond <span class="sc">&amp;</span> treated_cond]</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>gene1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "FBgn0033367"</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Identify gene that is expressed more in untreated samples</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>control_cond <span class="ot">&lt;-</span> control_stats<span class="sc">$</span>Mean <span class="sc">&gt;</span> <span class="dv">7</span> <span class="sc">&amp;</span> control_stats<span class="sc">$</span>Mean <span class="sc">&lt;</span> <span class="dv">8</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>treated_cond <span class="ot">&lt;-</span> treated_stats<span class="sc">$</span>Mean <span class="sc">&gt;</span> <span class="dv">4</span> <span class="sc">&amp;</span> treated_stats<span class="sc">$</span>Mean <span class="sc">&lt;</span> <span class="dv">5</span></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>gene2 <span class="ot">&lt;-</span> <span class="fu">rownames</span>(norm_cts)[control_cond <span class="sc">&amp;</span> treated_cond]</span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>gene2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "FBgn0039155"</code></pre>
</div>
</div>
<p>The expression of FBgn0039155 shows consistent differences between the control and treated conditions.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualise differential expression of selected gene</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(log_cts[gene2, ], <span class="at">ylab =</span> <span class="st">"Log Expression"</span>)</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a><span class="fu">title</span>(gene2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="plots/unnamed-chunk-21-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Our data fit the basic assumptions of DEseq, as it based on negative binomial distribution.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DESeq2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Run deseq2</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span>(<span class="at">countData =</span> cts,</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>                              <span class="at">colData =</span> coldata,</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>                              <span class="at">design =</span> <span class="sc">~</span> condition)</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeq</span>(dds)</span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">results</span>(dds)</span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a>res</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>log2 fold change (MLE): condition untreated vs treated 
Wald test p-value: condition untreated vs treated 
DataFrame with 14599 rows and 6 columns
               baseMean log2FoldChange     lfcSE        stat    pvalue
              &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt; &lt;numeric&gt;
FBgn0000003    0.171569    -1.02604541  3.805503 -0.26962147 0.7874515
FBgn0000008   95.144079    -0.00215142  0.223884 -0.00960955 0.9923328
FBgn0000014    1.056572     0.49673557  2.160264  0.22994204 0.8181368
FBgn0000015    0.846723     1.88276170  2.106432  0.89381546 0.3714206
FBgn0000017 4352.592899     0.24002523  0.126024  1.90459450 0.0568328
...                 ...            ...       ...         ...       ...
FBgn0261571 8.73437e-02     -0.9002942  3.810165  -0.2362875  0.813210
FBgn0261572 6.19714e+00      0.9591315  0.777017   1.2343759  0.217063
FBgn0261573 2.24098e+03     -0.0126158  0.112701  -0.1119412  0.910870
FBgn0261574 4.85774e+03     -0.0152569  0.193148  -0.0789905  0.937040
FBgn0261575 1.06836e+01     -0.1635594  0.938909  -0.1742016  0.861707
                 padj
            &lt;numeric&gt;
FBgn0000003        NA
FBgn0000008  0.996927
FBgn0000014        NA
FBgn0000015        NA
FBgn0000017  0.282361
...               ...
FBgn0261571        NA
FBgn0261572        NA
FBgn0261573  0.982035
FBgn0261574  0.988143
FBgn0261575  0.967912</code></pre>
</div>
</div>
<p>Curiously, FBgn0039155 was designated as the most differentially expressed gene by DESeq2 (effect size: 4.62, p-value: 4.07^{-161}), whereas FBgn0033367 did not appear among the top 10 genes but showed statistical significance (effect size: -1.45, p-value: 10^{-9}).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># View genes ordered by differential expression</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>sorted_res <span class="ot">&lt;-</span> res[<span class="fu">order</span>(res<span class="sc">$</span>padj), <span class="fu">c</span>(<span class="st">"log2FoldChange"</span>, <span class="st">"padj"</span>)]</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(sorted_res, <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>log2 fold change (MLE): condition untreated vs treated 
 
DataFrame with 10 rows and 2 columns
            log2FoldChange         padj
                 &lt;numeric&gt;    &lt;numeric&gt;
FBgn0039155        4.61901 4.06607e-161
FBgn0025111       -2.89986 6.38334e-112
FBgn0029167        2.19700 3.69146e-110
FBgn0003360        3.17967 1.98854e-105
FBgn0035085        2.56041  2.14348e-74
FBgn0039827        4.16252  1.74302e-68
FBgn0034736        3.51144  4.59386e-57
FBgn0029896        2.44502  3.03284e-55
FBgn0000071       -2.67958  8.79175e-46
FBgn0051092       -2.32771  1.09010e-36</code></pre>
</div>
</div>
</section>
<section id="part-5" class="level1">
<h1>Part 5</h1>
<p>The circadian RNA-Seq data from Zebrafish is imported and processed into a matrix with unique rownames.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Import dataset</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>circ_df <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"data/CircadianRNAseq.csv"</span>)</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Preprocess dataset</span></span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(circ_df) <span class="ot">&lt;-</span> <span class="fu">make.unique</span>(circ_df<span class="sc">$</span>GeneSymbol)</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>circ_df[ , <span class="fu">c</span>(<span class="st">"RefSeqID"</span>, <span class="st">"GeneSymbol"</span>)] <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>circ_mat <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(circ_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The last five rows (or genes) of the assay are printed below. From the colnames, it can be deduced that samples were collected from 11 PM until 7 PM for two consecutive days, with a time interval of 4 hours.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Show expression of last 5 genes</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a><span class="fu">tail</span>(circ_mat, <span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>              A_11PM     A_3AM     A_7AM    A_11AM     A_3PM     A_7PM
zgc:136896 255.99900 209.97600 125.42200 143.07200 112.50300 101.81700
LCP1         5.72658   6.62624   6.62286   9.62324   5.98339   5.37718
Rab3gap1.1  11.32120  14.44910  15.91110  11.89190  13.33890  12.65970
ACTR3       50.64750  56.36470  66.65160  85.98560  54.14030  59.21070
UBXN4       13.65380  15.55590  23.63540  18.67810  18.75290  16.04330
              B_11PM     B_3AM     B_7AM   B_11AM    B_3PM    B_7PM
zgc:136896 201.12300 118.10400 255.72500 89.71120 92.12400 86.27090
LCP1         3.90986   9.07169   6.51965  6.85435  5.85492  4.42235
Rab3gap1.1  10.15590  16.13480  13.85730 15.99760 15.21870 16.85700
ACTR3       53.61420  58.49680  51.49100 46.21930 50.31010 45.71560
UBXN4       13.86520  20.21520  15.92240 19.05500 19.76430 19.02300</code></pre>
</div>
</div>
<p>The expression of the per1a gene shows a rather consistent periodicity of 24 hours over the two days of sampling. Such observation agrees with the literature, as per1a is one of the genes responsible for the circadian rhythm of insects as well as mammals, where it is mostly expressed in the area of the brain known as suprachiasmatic nucleus (SPN).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualise oscillatory pattern of per1a gene over samples</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(circ_mat[<span class="st">"per1a"</span>, ], <span class="at">ylab =</span> <span class="st">"Norm Expression"</span>, <span class="at">type =</span> <span class="st">"b"</span>, <span class="at">xaxt =</span> <span class="st">"n"</span>)</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a><span class="fu">axis</span>(<span class="dv">1</span>, <span class="at">at =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">12</span>, <span class="at">labels =</span> <span class="fu">colnames</span>(circ_mat)) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="plots/unnamed-chunk-27-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Find frequency powers</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>powers <span class="ot">&lt;-</span> <span class="fu">fft</span>(circ_mat[<span class="st">"per1a"</span>, ])</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>powers <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(powers <span class="sc">*</span> <span class="fu">Conj</span>(powers))[<span class="dv">2</span><span class="sc">:</span>(<span class="dv">1</span> <span class="sc">+</span> <span class="fu">length</span>(powers) <span class="sc">/</span> <span class="dv">2</span>)]</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>norm_powers <span class="ot">&lt;-</span> powers <span class="sc">/</span> <span class="fu">sum</span>(powers)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Find base frequencies</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>t <span class="ot">&lt;-</span> <span class="dv">4</span></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>min_freq <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">/</span> (<span class="fu">ncol</span>(circ_mat) <span class="sc">*</span> t)</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>max_freq <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">/</span> (<span class="dv">2</span> <span class="sc">*</span> t)</span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>freq_range <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="at">from =</span> min_freq, <span class="at">to =</span> max_freq, <span class="at">by =</span> min_freq)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualise frequency components of oscillatory pattern for per1a</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(freq_range, norm_powers, <span class="at">xlab =</span> <span class="st">"Frequency in 1/h"</span>, <span class="at">ylab =</span> <span class="st">"Norm Power"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="plots/unnamed-chunk-30-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Between the options of (a) designing an experiment with (a) a higher time resolution or (b) a longer duration, we believe that option (b) is preferred, because, as we learned and applied at the beginning of the exercise, in order to get more reliable results, there is first of all a need to perform repeated measurements of the same situation. Therefore, if we want to prove that a certain gene operates at a circadian frequency, we must first measure it a sufficient number of times in order to prove that it is not an error/noise/non-normalized results and more. In addition, in order to show that this is not a temporary or passing phenomenon, measurement of multiple days will make it possible to prove that the gene works in a circadian manner regularly. However, option (a) can yield a more accurate estimate of the frequency components of the signal and to increase the Nyquist frequency, so it can be useful in situations where we want to get the best shape of the signal expressing the expression level or get a wider range of frequencies along the way.</p>
<p>Frequency analysis represents a more powerful tool than time analysis to detect the oscillatory nature of biological signals. Because signals are almost always composed by several waves with different sub-frequencies and mixed up with noise, it is not practical to study their oscillatory nature in the time domain. Instead, the frequency domain allows to separate the frequency components of the signal, which can be thereby described as the sum of simple waves, such as sine and cosine functions. In addition, working with frequencies you can take out noises and get better results for the circadian cycle.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Define function to obtain frequency powers with fft</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>compute_powers <span class="ot">&lt;-</span> <span class="cf">function</span>(time_series) {</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>  powers <span class="ot">&lt;-</span> <span class="fu">fft</span>(time_series)</span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>  powers <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(powers <span class="sc">*</span> <span class="fu">Conj</span>(powers))[<span class="dv">2</span><span class="sc">:</span>(<span class="dv">1</span> <span class="sc">+</span> <span class="fu">length</span>(powers) <span class="sc">/</span> <span class="dv">2</span>)]</span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>  norm_powers <span class="ot">&lt;-</span> powers <span class="sc">/</span> <span class="fu">sum</span>(powers)</span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(norm_powers)</span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Obtain frequency powers for every gene in assay</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>power_mat <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">apply</span>(circ_mat, <span class="dv">1</span>, compute_powers))</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Sort genes by their 24-hour frequency power</span></span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a>power_idx <span class="ot">&lt;-</span> <span class="fu">order</span>(power_mat[ , <span class="dv">2</span>], <span class="at">decreasing =</span> <span class="cn">TRUE</span>)</span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a>sorted_powers <span class="ot">&lt;-</span> <span class="fu">rownames</span>(power_mat)[power_idx]</span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a><span class="co"># sort provides an easier way</span></span>
<span id="cb42-9"><a href="#cb42-9" aria-hidden="true" tabindex="-1"></a><span class="co"># sorted_powers &lt;- sort(power_mat[ , 2], decreasing = TRUE)</span></span>
<span id="cb42-10"><a href="#cb42-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-11"><a href="#cb42-11" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(sorted_powers, <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "atxn1b"    "fus"       "nr1d2b"    "rdh1l"     "ankrd10a"  "phyhd1"   
 [7] "arntl1b.1" "ARNTL2"    "Ldhd"      "aclya"    </code></pre>
</div>
</div>
<p>We found out that one of the relevant genes - ‘nr1d2b’, is involved in regulation of circadian rhythm.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check if na values are present</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">is.na</span>(sorted_powers))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0</code></pre>
</div>
</div>
</section>
<section id="part-6" class="level1">
<h1>Part 6</h1>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute log statistics of each gene with custom function from part 3</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>circ_stats <span class="ot">&lt;-</span> <span class="fu">find_log_stats</span>(circ_mat, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The genes are binned by similarity in library size and a z-score is calculated within each bin. The top 40 genes with the highest z-score are then inspected to understand their relation with circadian rhythms.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tibble)</span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Bin genes and compute their z-score</span></span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>zscore_df <span class="ot">&lt;-</span> circ_stats <span class="sc">%&gt;%</span></span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="st">"GeneSymbol"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(Mean <span class="sc">&gt;=</span> <span class="dv">3</span>) <span class="sc">%&gt;%</span></span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">Bin =</span> <span class="fu">ntile</span>(Mean, <span class="dv">20</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb47-9"><a href="#cb47-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(Bin) <span class="sc">%&gt;%</span></span>
<span id="cb47-10"><a href="#cb47-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">ZScore =</span> (Var <span class="sc">-</span> <span class="fu">mean</span>(Var)) <span class="sc">/</span> <span class="fu">sd</span>(Var)) <span class="sc">%&gt;%</span></span>
<span id="cb47-11"><a href="#cb47-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(ZScore))</span>
<span id="cb47-12"><a href="#cb47-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-13"><a href="#cb47-13" aria-hidden="true" tabindex="-1"></a><span class="co"># View genes sorted by z-score</span></span>
<span id="cb47-14"><a href="#cb47-14" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(zscore_df, <span class="dv">40</span>), <span class="at">digits =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 67%">
<col style="width: 8%">
<col style="width: 7%">
<col style="width: 5%">
<col style="width: 10%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">GeneSymbol</th>
<th style="text-align: right;">Var</th>
<th style="text-align: right;">Mean</th>
<th style="text-align: right;">Bin</th>
<th style="text-align: right;">ZScore</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">CS.2</td>
<td style="text-align: right;">9.77</td>
<td style="text-align: right;">3.67</td>
<td style="text-align: right;">11</td>
<td style="text-align: right;">4.82</td>
</tr>
<tr class="even">
<td style="text-align: left;">LOC792433</td>
<td style="text-align: right;">7.41</td>
<td style="text-align: right;">3.09</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">4.45</td>
</tr>
<tr class="odd">
<td style="text-align: left;">rho</td>
<td style="text-align: right;">11.62</td>
<td style="text-align: right;">4.90</td>
<td style="text-align: right;">18</td>
<td style="text-align: right;">3.75</td>
</tr>
<tr class="even">
<td style="text-align: left;">exorh.1</td>
<td style="text-align: right;">16.50</td>
<td style="text-align: right;">9.71</td>
<td style="text-align: right;">20</td>
<td style="text-align: right;">3.72</td>
</tr>
<tr class="odd">
<td style="text-align: left;">zgc:136930</td>
<td style="text-align: right;">7.83</td>
<td style="text-align: right;">3.59</td>
<td style="text-align: right;">10</td>
<td style="text-align: right;">3.57</td>
</tr>
<tr class="even">
<td style="text-align: left;">zgc:194737</td>
<td style="text-align: right;">6.13</td>
<td style="text-align: right;">3.02</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">3.51</td>
</tr>
<tr class="odd">
<td style="text-align: left;">.469</td>
<td style="text-align: right;">6.77</td>
<td style="text-align: right;">3.14</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">3.47</td>
</tr>
<tr class="even">
<td style="text-align: left;">si:ch211-132b12.7.1</td>
<td style="text-align: right;">8.09</td>
<td style="text-align: right;">3.67</td>
<td style="text-align: right;">11</td>
<td style="text-align: right;">3.43</td>
</tr>
<tr class="odd">
<td style="text-align: left;">zgc:92061.1</td>
<td style="text-align: right;">7.97</td>
<td style="text-align: right;">3.54</td>
<td style="text-align: right;">9</td>
<td style="text-align: right;">3.35</td>
</tr>
<tr class="even">
<td style="text-align: left;">zgc:103748</td>
<td style="text-align: right;">6.88</td>
<td style="text-align: right;">3.36</td>
<td style="text-align: right;">7</td>
<td style="text-align: right;">3.32</td>
</tr>
<tr class="odd">
<td style="text-align: left;">olfm1b.1</td>
<td style="text-align: right;">6.22</td>
<td style="text-align: right;">3.26</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">3.25</td>
</tr>
<tr class="even">
<td style="text-align: left;">fabp7b</td>
<td style="text-align: right;">7.76</td>
<td style="text-align: right;">3.52</td>
<td style="text-align: right;">9</td>
<td style="text-align: right;">3.18</td>
</tr>
<tr class="odd">
<td style="text-align: left;">cldna</td>
<td style="text-align: right;">6.14</td>
<td style="text-align: right;">3.22</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">3.17</td>
</tr>
<tr class="even">
<td style="text-align: left;">regulation of transcription, DNA-dependent .1</td>
<td style="text-align: right;">5.77</td>
<td style="text-align: right;">3.01</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">3.16</td>
</tr>
<tr class="odd">
<td style="text-align: left;">.120</td>
<td style="text-align: right;">8.30</td>
<td style="text-align: right;">4.00</td>
<td style="text-align: right;">14</td>
<td style="text-align: right;">3.13</td>
</tr>
<tr class="even">
<td style="text-align: left;">nr1d2b.1</td>
<td style="text-align: right;">7.30</td>
<td style="text-align: right;">3.56</td>
<td style="text-align: right;">10</td>
<td style="text-align: right;">3.10</td>
</tr>
<tr class="odd">
<td style="text-align: left;">bhlhb3l</td>
<td style="text-align: right;">8.79</td>
<td style="text-align: right;">4.06</td>
<td style="text-align: right;">15</td>
<td style="text-align: right;">3.09</td>
</tr>
<tr class="even">
<td style="text-align: left;">.161</td>
<td style="text-align: right;">7.74</td>
<td style="text-align: right;">3.82</td>
<td style="text-align: right;">13</td>
<td style="text-align: right;">3.06</td>
</tr>
<tr class="odd">
<td style="text-align: left;">aqp3a</td>
<td style="text-align: right;">8.69</td>
<td style="text-align: right;">4.19</td>
<td style="text-align: right;">15</td>
<td style="text-align: right;">3.01</td>
</tr>
<tr class="even">
<td style="text-align: left;">FOS</td>
<td style="text-align: right;">6.62</td>
<td style="text-align: right;">3.32</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">3.01</td>
</tr>
<tr class="odd">
<td style="text-align: left;">scinlb</td>
<td style="text-align: right;">6.55</td>
<td style="text-align: right;">3.38</td>
<td style="text-align: right;">7</td>
<td style="text-align: right;">3.01</td>
</tr>
<tr class="even">
<td style="text-align: left;">cfdl</td>
<td style="text-align: right;">7.66</td>
<td style="text-align: right;">3.90</td>
<td style="text-align: right;">13</td>
<td style="text-align: right;">2.99</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Cox7c</td>
<td style="text-align: right;">8.77</td>
<td style="text-align: right;">4.36</td>
<td style="text-align: right;">16</td>
<td style="text-align: right;">2.96</td>
</tr>
<tr class="even">
<td style="text-align: left;">CFL1</td>
<td style="text-align: right;">8.08</td>
<td style="text-align: right;">3.98</td>
<td style="text-align: right;">14</td>
<td style="text-align: right;">2.94</td>
</tr>
<tr class="odd">
<td style="text-align: left;">zgc:162144</td>
<td style="text-align: right;">7.07</td>
<td style="text-align: right;">3.57</td>
<td style="text-align: right;">10</td>
<td style="text-align: right;">2.90</td>
</tr>
<tr class="even">
<td style="text-align: left;">Camk2n1</td>
<td style="text-align: right;">5.84</td>
<td style="text-align: right;">3.26</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">2.87</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Cyp11b2</td>
<td style="text-align: right;">5.94</td>
<td style="text-align: right;">3.17</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">2.86</td>
</tr>
<tr class="even">
<td style="text-align: left;">arg2</td>
<td style="text-align: right;">5.90</td>
<td style="text-align: right;">3.15</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">2.83</td>
</tr>
<tr class="odd">
<td style="text-align: left;">icn2</td>
<td style="text-align: right;">6.94</td>
<td style="text-align: right;">3.61</td>
<td style="text-align: right;">10</td>
<td style="text-align: right;">2.79</td>
</tr>
<tr class="even">
<td style="text-align: left;">ndrg1l</td>
<td style="text-align: right;">9.38</td>
<td style="text-align: right;">4.58</td>
<td style="text-align: right;">17</td>
<td style="text-align: right;">2.79</td>
</tr>
<tr class="odd">
<td style="text-align: left;">.241</td>
<td style="text-align: right;">5.65</td>
<td style="text-align: right;">3.04</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">2.78</td>
</tr>
<tr class="even">
<td style="text-align: left;">cldnb</td>
<td style="text-align: right;">6.78</td>
<td style="text-align: right;">3.43</td>
<td style="text-align: right;">8</td>
<td style="text-align: right;">2.77</td>
</tr>
<tr class="odd">
<td style="text-align: left;">UQCRQ</td>
<td style="text-align: right;">7.08</td>
<td style="text-align: right;">3.74</td>
<td style="text-align: right;">12</td>
<td style="text-align: right;">2.74</td>
</tr>
<tr class="even">
<td style="text-align: left;">per1b</td>
<td style="text-align: right;">7.25</td>
<td style="text-align: right;">3.66</td>
<td style="text-align: right;">11</td>
<td style="text-align: right;">2.74</td>
</tr>
<tr class="odd">
<td style="text-align: left;">zgc:63920.1</td>
<td style="text-align: right;">5.58</td>
<td style="text-align: right;">3.05</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">2.72</td>
</tr>
<tr class="even">
<td style="text-align: left;">Cry3</td>
<td style="text-align: right;">5.91</td>
<td style="text-align: right;">3.14</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">2.71</td>
</tr>
<tr class="odd">
<td style="text-align: left;">gria2b</td>
<td style="text-align: right;">6.26</td>
<td style="text-align: right;">3.29</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">2.69</td>
</tr>
<tr class="even">
<td style="text-align: left;">flr.1</td>
<td style="text-align: right;">5.65</td>
<td style="text-align: right;">3.21</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">2.68</td>
</tr>
<tr class="odd">
<td style="text-align: left;">si:dkey-7c18.24</td>
<td style="text-align: right;">8.27</td>
<td style="text-align: right;">4.19</td>
<td style="text-align: right;">15</td>
<td style="text-align: right;">2.67</td>
</tr>
<tr class="even">
<td style="text-align: left;">LOC563601</td>
<td style="text-align: right;">5.86</td>
<td style="text-align: right;">3.12</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">2.66</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>We can detect some known circadian genes from this table: For example – the gene ‘nr1d2b’ is involved in regulation of circadian rhythm. In addition, the gene ‘bhlhb3l’ is involved in regulation of circadian rhythms. It Is expected that circadian genes will also be variable gene, as they are changing over time during the day, so it make sense that they have big differences in expression levels. Also, changes in environmental conditions throughout the day may lead to the need for the body to adapt to different conditions such as different weather during the day or night or light and darkness, therefore the ability to change in cycles over time can lead to increased variability in expression</p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>